Release/src/http/client/http_client_winhttp.cpp

```diff
@@ -1100,136 +1028,88 @@ class winhttp_client : public _http_client_communicator
+                DWORD urlSize{ 0 };
+                WinHttpQueryOption(hRequestHandle, WINHTTP_OPTION_URL, NULL, &urlSize);
+                auto urlwchar = new WCHAR[urlSize / sizeof(WCHAR)];
+
+                WinHttpQueryOption(hRequestHandle, WINHTTP_OPTION_URL, (void*)urlwchar, &urlSize);
+                utility::string_t url(urlwchar);
+
+                delete[] urlwchar;
+
+                // obtain leaf cert based on which we will be able to build the certificate chain
+                PCCERT_CONTEXT pCert{ nullptr };
+                DWORD dwSize = sizeof(pCert);
+
+                WinHttpQueryOption(hRequestHandle, WINHTTP_OPTION_SERVER_CERT_CONTEXT, &pCert, &dwSize);
+
+                std::vector<std::pair<const unsigned char*, int>> cert_chain;
+
+                if (pCert)
+                {
+                    CERT_ENHKEY_USAGE keyUsage = {};
+                    keyUsage.cUsageIdentifier = 0;
+                    keyUsage.rgpszUsageIdentifier = NULL;
+
+                    CERT_USAGE_MATCH certUsage = {};
+                    certUsage.dwType = USAGE_MATCH_TYPE_AND;
+                    certUsage.Usage = keyUsage;
+
+                    CERT_CHAIN_PARA chainPara = {};
+                    chainPara.cbSize = sizeof(CERT_CHAIN_PARA);
+                    chainPara.RequestedUsage = certUsage;
+
+                    PCCERT_CHAIN_CONTEXT pChainContext = {};
+
+                    // build the certificate chain relying on the actual intermediate certs returned as part of the TLS session
+                    // disable any network operations to fetch certificates
+                    auto validChain = CertGetCertificateChain(NULL, pCert, NULL, pCert->hCertStore, &chainPara,
+                        CERT_CHAIN_CACHE_ONLY_URL_RETRIEVAL |
+                        CERT_CHAIN_REVOCATION_CHECK_CACHE_ONLY |
+                        CERT_CHAIN_REVOCATION_CHECK_CHAIN |
+                        CERT_CHAIN_CACHE_END_CERT |
+                        CERT_CHAIN_DISABLE_AUTH_ROOT_AUTO_UPDATE,
+                        NULL, &pChainContext);
+
+                    if (validChain && pChainContext)
+                    {
+                        // extract all certificates from the the TLS connection
+                        cert_chain.reserve((int)pChainContext->cChain);
+                        for (size_t i = 0; i < pChainContext->cChain; ++i)
+                        {
+                            auto chain = pChainContext->rgpChain[i];
+                            for (size_t j = 0; j < chain->cElement; ++j)
+                            {
+                                auto chainElement = chain->rgpElement[j];
+                                auto cert = chainElement->pCertContext;
+                                if (cert)
+                                {
+                                    cert_chain.emplace_back(cert->pbCertEncoded, (int)cert->cbCertEncoded);
```

Comment:
So the lifecycle of pbCertEncoded is tied to the lifecycle of pChainContext, and the int is not the length, but a bool!